File: packages/types/src/signing.ts
/**
 * HMAC-SHA256 signing helpers for Canvas Kit v2.
 *
 * Renowide signs every outbound request to a developer's canvas host or
 * action webhook with HMAC-SHA256. The developer verifies the signature
 * with the same secret (`handoff_secret`, stored on Renowide as
 * `agent_profiles.webhook_secret`).
 *
 * Two canonical strings — DO NOT invent new ones. If you add a new
 * surface (e.g. state-stream SSE), we extend this file with a new
 * `sign<Surface>Request` function and a new version prefix.
 *
 *   GET  /canvas/<surface>[.json]                   ← `v1:<ts>:<slug>:<surface>:<buyer|->:<hire|->:<rid>`
 *   POST /action                                    ← `v1:<ts>:<raw-body-bytes>`
 *
 * Wire format: `Renowide-Signature: v1=<hex>` (lowercase hex). The
 * developer rejects any request where the signature doesn't match or
 * where |now - ts| > 300 seconds.
 *
 * All helpers here are environment-agnostic: they accept `createHmac` from
 * `node:crypto` by default, but you can inject a web-crypto polyfill if
 * you're running in a worker / edge runtime.
 */

import { createHmac, timingSafeEqual as nodeTimingSafeEqual } from "node:crypto";

export const SIGNATURE_SCHEME_VERSION = "v1" as const;
export const SIGNATURE_MAX_CLOCK_SKEW_SECONDS = 300;

export interface SignCanvasRequestArgs {
  handoffSecret: string;
  agentSlug: string;
  surface: "hire_flow" | "post_hire";
  buyerId: string | null | undefined;
  hireId: string | null | undefined;
  requestId: string;
  timestamp: number;
}

/**
 * Sign a canvas-fetch GET request. The canonical string is:
 *
 *     v1:<timestamp>:<agent_slug>:<surface>:<buyer_id|->:<hire_id|->:<request_id>
 *
 * Missing buyer/hire ids are rendered as `-`.
 *
 * Returns the 64-char lowercase-hex HMAC digest. Wire it into the
 * `Renowide-Signature: v1=<hex>` header on the request.
 */
export function signCanvasRequest(args: SignCanvasRequestArgs): string {
  const canonical =
    `${SIGNATURE_SCHEME_VERSION}:${args.timestamp}:${args.agentSlug}:${args.surface}:` +
    `${args.buyerId ?? "-"}:${args.hireId ?? "-"}:${args.requestId}`;
  const mac = createHmac("sha256", args.handoffSecret);
  mac.update(canonical);
  return mac.digest("hex");
}

/**
 * Sign a canvas-action POST request. The canonical string is:
 *
 *     v1:<timestamp>:<raw-body-bytes>
 *
 * `body` must be the **exact bytes** that go on the wire — never the
 * parsed JSON (re-serialising may reorder keys and break the signature).
 * Most HTTP middleware exposes the raw body via `req.rawBody`.
 */
export function signActionRequest(
  handoffSecret: string,
  body: Uint8Array,
  timestamp: number,
): string {
  const prefix = Buffer.from(`${SIGNATURE_SCHEME_VERSION}:${timestamp}:`, "utf-8");
  const mac = createHmac("sha256", handoffSecret);
  mac.update(prefix);
  mac.update(body);
  return mac.digest("hex");
}

// ─── Verification helpers (for developer webhook handlers) ─────────────────

export class SignatureVerificationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "SignatureVerificationError";
  }
}

export interface VerifyCanvasRequestArgs {
  handoffSecret: string;
  headers: Record<string, string | undefined>;
  agentSlug: string;
  surface: "hire_flow" | "post_hire";
  /** Usually `Date.now() / 1000` floored. Override in tests. */
  nowSeconds?: number;
  /** Optional skew override (default 300s). */
  maxClockSkewSeconds?: number;
}

/**
 * Verify an inbound canvas-fetch GET. Throws `SignatureVerificationError`
 * on failure. Call this from your `/canvas/hire_flow.json` + `/canvas/post_hire/...` handlers.
 */
export function verifyCanvasRequest(args: VerifyCanvasRequestArgs): void {
  const headers = lowercaseHeaders(args.headers);
  const scheme = parseSignatureHeader(headers["renowide-signature"]);
  const tsRaw = headers["x-renowide-timestamp"];
  const requestId = headers["x-renowide-request-id"];
  const buyerId = headers["x-renowide-buyer-id"];
  const hireId = headers["x-renowide-hire-id"];

  if (!tsRaw) throw new SignatureVerificationError("missing X-Renowide-Timestamp");
  if (!requestId) throw new SignatureVerificationError("missing X-Renowide-Request-Id");

  checkClockSkew(
    parseIntSafe(tsRaw),
    args.nowSeconds ?? Math.floor(Date.now() / 1000),
    args.maxClockSkewSeconds ?? SIGNATURE_MAX_CLOCK_SKEW_SECONDS,
  );

  const expected = signCanvasRequest({
    handoffSecret: args.handoffSecret,
    agentSlug: args.agentSlug,
    surface: args.surface,
    buyerId: buyerId ?? null,
    hireId: hireId ?? null,
    requestId,
    timestamp: parseIntSafe(tsRaw),
  });

  if (!constantTimeHexEqual(scheme.hex, expected)) {
    throw new SignatureVerificationError("signature mismatch");
  }
}

export interface VerifyActionRequestArgs {
  handoffSecret: string;
  headers: Record<string, string | undefined>;
  /** Raw request body bytes — NOT the parsed JSON. */
  body: Uint8Array;
  nowSeconds?: number;
  maxClockSkewSeconds?: number;
}

/**
 * Verify an inbound action POST. Throws `SignatureVerificationError` on
 * failure. Call this from your `action_webhook_url` handler before you
 * JSON.parse the body.
 */
export function verifyActionRequest(args: VerifyActionRequestArgs): void {
  const headers = lowercaseHeaders(args.headers);
  const scheme = parseSignatureHeader(headers["renowide-signature"]);
  const tsRaw = headers["x-renowide-timestamp"];
  if (!tsRaw) throw new SignatureVerificationError("missing X-Renowide-Timestamp");

  checkClockSkew(
    parseIntSafe(tsRaw),
    args.nowSeconds ?? Math.floor(Date.now() / 1000),
    args.maxClockSkewSeconds ?? SIGNATURE_MAX_CLOCK_SKEW_SECONDS,
  );

  const expected = signActionRequest(args.handoffSecret, args.body, parseIntSafe(tsRaw));
  if (!constantTimeHexEqual(scheme.hex, expected)) {
    throw new SignatureVerificationError("signature mismatch");
  }
}

// ─── Private helpers ───────────────────────────────────────────────────────

/**
 * Case-insensitive header lookup — mirrors
 * `renowide_canvas.signing._get_header` in the Python sibling. Any
 * duplicate header names (which shouldn't happen for signing headers
 * but we don't trust the caller) resolve to the *last* casing — same
 * semantics as iterating a case-insensitive `Mapping.items()` in
 * Starlette / Flask.
 */
function lowercaseHeaders(
  headers: Record<string, string | undefined>,
): Record<string, string | undefined> {
  const out: Record<string, string | undefined> = Object.create(null);
  for (const [k, v] of Object.entries(headers)) {
    out[k.toLowerCase()] = v;
  }
  return out;
}

function parseSignatureHeader(raw: string | undefined): { version: string; hex: string } {
  if (!raw) throw new SignatureVerificationError("missing Renowide-Signature");
  // `split("=", 1)` in Python returns *at most 2* parts and puts the
  // remainder into the second — the opposite of JS's `split(sep, limit)`
  // which *truncates*. Mirror Python exactly by splitting on the FIRST
  // `=` only, so a tampered `v1=abcd=extra` header behaves the same way
  // on both verifiers (rejected by the hex-charset check below).
  const eq = raw.indexOf("=");
  const version = eq === -1 ? raw : raw.slice(0, eq);
  const hex = eq === -1 ? "" : raw.slice(eq + 1);
  if (!version || !hex || version !== SIGNATURE_SCHEME_VERSION) {
    throw new SignatureVerificationError(
      `unsupported signature scheme (expected ${SIGNATURE_SCHEME_VERSION}=<hex>, got ${raw})`,
    );
  }
  if (!/^[0-9a-f]{64}$/.test(hex)) {
    throw new SignatureVerificationError("signature is not 64 hex chars");
  }
  return { version, hex };
}

/**
 * Strict integer parser — mirrors Python's `int(raw)`. Unlike
 * `Number.parseInt("1234abc", 10)` which accepts trailing garbage, this
 * helper rejects any non-digit (other than a single leading `-`). That
 * closes the cross-language divergence where a tampered
 * `X-Renowide-Timestamp: 1700000000abc` passes the TS verifier but is
 * rejected by the Python one.
 */
function parseIntSafe(v: string): number {
  if (!/^-?\d+$/.test(v.trim())) {
    throw new SignatureVerificationError(`bad integer header: ${v}`);
  }
  const n = Number.parseInt(v.trim(), 10);
  if (!Number.isFinite(n)) {
    throw new SignatureVerificationError(`bad integer header: ${v}`);
  }
  return n;
}

function checkClockSkew(requestTs: number, nowSeconds: number, max: number): void {
  if (Math.abs(nowSeconds - requestTs) > max) {
    throw new SignatureVerificationError(
      `clock skew too large (|${nowSeconds} - ${requestTs}| > ${max}s)`,
    );
  }
}

function constantTimeHexEqual(a: string, b: string): boolean {
  if (a.length !== b.length) return false;
  const bufA = Buffer.from(a, "hex");
  const bufB = Buffer.from(b, "hex");
  if (bufA.length !== bufB.length) return false;
  return nodeTimingSafeEqual(bufA, bufB);
}

File: python/renowide-canvas/src/renowide_canvas/signing.py
"""HMAC-SHA256 signing + verification for Canvas Kit v2.

Mirror of ``@renowide/types/signing`` in the JavaScript monorepo. The
canonical strings and wire headers must stay bit-identical across the
two implementations; if you change one, change the other.

Canonical strings
-----------------

* Canvas GET (``/canvas/<surface>.json``)::

      v1:<ts>:<agent_slug>:<surface>:<buyer_id|->:<hire_id|->:<request_id>

* Action POST (``/actions``)::

      v1:<ts>:<raw_body_bytes>

  where *raw_body_bytes* is the exact on-the-wire bytes of the JSON
  payload — not a re-serialisation.

Wire header::

    Renowide-Signature: v1=<64-char lowercase hex>

Additional headers on every inbound request:

* ``X-Renowide-Timestamp``   — Unix seconds integer.
* ``X-Renowide-Request-Id``  — ULID.
* ``X-Renowide-Buyer-Id``    — optional.
* ``X-Renowide-Hire-Id``     — optional.

We reject stale requests (``|now - ts| > 300 s``).
"""

from __future__ import annotations

import hashlib
import hmac
import time
from dataclasses import dataclass
from typing import Mapping, Optional, Union

SIGNATURE_SCHEME_VERSION: str = "v1"
SIGNATURE_MAX_CLOCK_SKEW_SECONDS: int = 300

__all__ = [
    "SIGNATURE_SCHEME_VERSION",
    "SIGNATURE_MAX_CLOCK_SKEW_SECONDS",
    "SignatureVerificationError",
    "sign_canvas_request",
    "sign_action_request",
    "verify_canvas_request",
    "verify_action_request",
]


class SignatureVerificationError(Exception):
    """Raised when a signed Renowide request fails verification.

    The ``code`` attribute is a short machine-readable string (see the
    SDK docs) so you can map it into a structured HTTP response.
    """

    def __init__(self, code: str, message: Optional[str] = None) -> None:
        super().__init__(message or code)
        self.code = code


# ─── Signing ────────────────────────────────────────────────────────────────


def sign_canvas_request(
    *,
    handoff_secret: Union[str, bytes],
    agent_slug: str,
    surface: str,
    buyer_id: Optional[str],
    hire_id: Optional[str],
    request_id: str,
    timestamp: int,
) -> str:
    """Return the lowercase-hex HMAC for a canvas GET request.

    Wire as ``Renowide-Signature: v1=<result>`` on the outbound
    request (or verify against it on the inbound side via
    :func:`verify_canvas_request`).
    """
    canonical = (
        f"{SIGNATURE_SCHEME_VERSION}:{timestamp}:{agent_slug}:{surface}:"
        f"{buyer_id or '-'}:{hire_id or '-'}:{request_id}"
    )
    key = handoff_secret.encode("utf-8") if isinstance(handoff_secret, str) else handoff_secret
    return hmac.new(key, canonical.encode("utf-8"), hashlib.sha256).hexdigest()


def sign_action_request(
    handoff_secret: Union[str, bytes],
    body: bytes,
    timestamp: int,
) -> str:
    """Return the lowercase-hex HMAC for an action POST request.

    ``body`` must be the **exact bytes** that will hit the wire; never
    re-serialised JSON (key ordering is not guaranteed stable).
    """
    if not isinstance(body, (bytes, bytearray, memoryview)):  # pragma: no cover - defensive
        raise TypeError("body must be bytes; never a parsed JSON object")

    key = handoff_secret.encode("utf-8") if isinstance(handoff_secret, str) else handoff_secret
    mac = hmac.new(key, None, hashlib.sha256)
    mac.update(f"{SIGNATURE_SCHEME_VERSION}:{timestamp}:".encode("utf-8"))
    mac.update(bytes(body))
    return mac.hexdigest()


# ─── Verification ───────────────────────────────────────────────────────────


@dataclass(frozen=True)
class _ParsedSigHeader:
    version: str
    hex_digest: str


def verify_canvas_request(
    *,
    handoff_secret: Union[str, bytes],
    headers: Mapping[str, str],
    agent_slug: str,
    surface: str,
    now_seconds: Optional[int] = None,
    max_clock_skew_seconds: int = SIGNATURE_MAX_CLOCK_SKEW_SECONDS,
) -> None:
    """Verify an inbound canvas-fetch GET.

    Raises :class:`SignatureVerificationError` on failure.

    ``headers`` should be the request's canonical header mapping
    (Starlette's ``request.headers`` works; Flask's ``request.headers``
    works; a plain dict works). Header name matching is case-insensitive.
    """
    sig = _parse_signature_header(_get_header(headers, "renowide-signature"))
    ts = _parse_int_header(_get_header(headers, "x-renowide-timestamp"), "x-renowide-timestamp")
    request_id = _get_header(headers, "x-renowide-request-id")
    if not request_id:
        raise SignatureVerificationError("missing_header", "x-renowide-request-id is required")

    buyer_id = _get_header(headers, "x-renowide-buyer-id") or None
    hire_id = _get_header(headers, "x-renowide-hire-id") or None

    _check_clock_skew(ts, now_seconds, max_clock_skew_seconds)

    expected = sign_canvas_request(
        handoff_secret=handoff_secret,
        agent_slug=agent_slug,
        surface=surface,
        buyer_id=buyer_id,
        hire_id=hire_id,
        request_id=request_id,
        timestamp=ts,
    )
    if not hmac.compare_digest(sig.hex_digest, expected):
        raise SignatureVerificationError("bad_signature", "HMAC mismatch")


def verify_action_request(
    *,
    handoff_secret: Union[str, bytes],
    headers: Mapping[str, str],
    body: bytes,
    now_seconds: Optional[int] = None,
    max_clock_skew_seconds: int = SIGNATURE_MAX_CLOCK_SKEW_SECONDS,
) -> None:
    """Verify an inbound action POST.

    ``body`` must be the raw request bytes (e.g. ``await request.body()``
    in FastAPI, ``request.get_data()`` in Flask). Do **not** pass
    ``request.json`` — re-serialisation will break the signature.
    """
    sig = _parse_signature_header(_get_header(headers, "renowide-signature"))
    ts = _parse_int_header(_get_header(headers, "x-renowide-timestamp"), "x-renowide-timestamp")
    _check_clock_skew(ts, now_seconds, max_clock_skew_seconds)

    expected = sign_action_request(handoff_secret, body, ts)
    if not hmac.compare_digest(sig.hex_digest, expected):
        raise SignatureVerificationError("bad_signature", "HMAC mismatch")


# ─── Internals ──────────────────────────────────────────────────────────────


def _get_header(headers: Mapping[str, str], name: str) -> Optional[str]:
    lower = name.lower()
    for k, v in headers.items():
        if k.lower() == lower:
            return v
    return None


def _parse_signature_header(raw: Optional[str]) -> _ParsedSigHeader:
    if not raw:
        raise SignatureVerificationError("missing_header", "Renowide-Signature header is required")
    parts = raw.split("=", 1)
    if len(parts) != 2:
        raise SignatureVerificationError("malformed_header", f"could not parse {raw!r}")
    version, hex_digest = parts[0].strip(), parts[1].strip().lower()
    if version != SIGNATURE_SCHEME_VERSION:
        raise SignatureVerificationError(
            "unsupported_version",
            f"unsupported signature scheme version {version!r}",
        )
    if len(hex_digest) != 64 or any(c not in "0123456789abcdef" for c in hex_digest):
        raise SignatureVerificationError(
            "malformed_header",
            "signature must be a 64-char lowercase hex digest",
        )
    return _ParsedSigHeader(version=version, hex_digest=hex_digest)


def _parse_int_header(raw: Optional[str], name: str) -> int:
    if not raw:
        raise SignatureVerificationError("missing_header", f"{name} is required")
    try:
        return int(raw)
    except ValueError:
        raise SignatureVerificationError("malformed_header", f"{name} must be an integer")


def _check_clock_skew(ts: int, now_seconds: Optional[int], max_skew: int) -> None:
    now = time.time() if now_seconds is None else now_seconds
    if abs(int(now) - ts) > max_skew:
        raise SignatureVerificationError(
            "stale_timestamp",
            f"timestamp skew {abs(int(now) - ts)}s exceeds max {max_skew}s",
        )
